pkg/client/client_v2.go

```diff
@@ -300,130 +302,129 @@ func (c *clientV2) startCheckin() {
 
 	go func() {
 		defer c.wg.Done()
-		for {
+		for c.ctx.Err() == nil {
+			// If the initial RPC connection fails, checkinRoundTrip
+			// returns immediately, so we set a timer to avoid spinlocking
+			// on the error.
+			retryTimer := time.NewTimer(time.Second)
+			c.checkinRoundTrip()
+
+			// After the RPC client closes, wait until either the timer interval
+			// expires or the context is cancelled. Note that the timer waits
+			// one second since the checkinRoundTrip was last _called_, not
+			// since it returns; immediate retries are ok after an active
+			// connection shuts down.
 			select {
+			case <-retryTimer.C:
 			case <-c.ctx.Done():
-				// stopped
-				return
-			default:
 			}
-
-			c.checkinRoundTrip()
 		}
 	}()
 }
 
+func (c *clientV2) checkinWriter(
+	checkinClient proto.ElasticAgent_CheckinV2Client,
+	done chan struct{},
+) {
+	t := time.NewTicker(c.minCheckTimeout)
+	defer t.Stop()
+
+	// Keep sending until the call returns an error
+	for c.sendObserved(checkinClient) == nil {
```

Comment:
Another option that doesn't rely on us manually wrapping each call is to look at using gRPC [interceptors](https://github.com/grpc/grpc-go/blob/master/examples/features/interceptor/README.md#interceptor) to do this to inject the retries into each underlying gRPC call.

Here's an implementation of a retry interceptor I've used before with linear and exponential backoff support:
https://pkg.go.dev/github.com/grpc-ecosystem/go-grpc-middleware/v2/interceptors/retry

Reply:
This sounds nice, but I just tried it and it doesn't seem to be working -- I've pushed my attempt so you can see it and suggest if there's anything I missed, but running it produces:

```
--- FAIL: TestRPCErrorRetryTimer (0.05s)
    client_v2_test.go:73:
        	Error Trace:	client_v2_test.go:73
        	Error:      	Not equal:
        	            	expected: 2
        	            	actual  : 11667
        	Test:       	TestRPCErrorRetryTimer
        	Messages:   	expected 2 RPC errors, one from checkinRoundTrip and one from actionRoundTrip
```

I tried a few different ways of setting it and can't find anything that works, so I think my theory is that this wraps retries of _calls_ to the RPC host, not the initial connection, which is what's causing the bug. (This fits with the interceptor docs saying it wraps "around the user call" which plausibly wouldn't include the creation of the client itself.) Let me know if you have other ideas, otherwise I can roll back to the previous implementation.

Reply:
Ah yes, it is retrying the individual calls and not the initial connection.